src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadataFilesCache.h

```diff
@@ -0,0 +1,165 @@
+         for (const auto & entry: manifest_file_cache_keys)
+         {
+             total_size += sizeof(Iceberg::ManifestFileEntry) + entry.manifest_file_path.capacity();
+         }
+         return total_size;
+    }
+
+    explicit IcebergMetadataFilesCacheCell(Poco::JSON::Object::Ptr metadata_object_, size_t memory_bytes_)
+        : cached_element(metadata_object_)
+        , memory_bytes(memory_bytes_ + SIZE_IN_MEMORY_OVERHEAD)
+        , metric_increment{CurrentMetrics::IcebergMetadataFilesCacheSize, memory_bytes}
+    {
+    }
+    explicit IcebergMetadataFilesCacheCell(ManifestFileCacheKeys && manifest_file_cache_keys_)
+        : cached_element(std::move(manifest_file_cache_keys_))
+        , memory_bytes(getMemorySizeOfManifestCacheKeys(std::get<ManifestFileCacheKeys>(cached_element)) + SIZE_IN_MEMORY_OVERHEAD)
+        , metric_increment{CurrentMetrics::IcebergMetadataFilesCacheSize, memory_bytes}
+    {
+    }
+    explicit IcebergMetadataFilesCacheCell(Iceberg::ManifestFilePtr manifest_file_)
+        : cached_element(manifest_file_)
+        , memory_bytes(std::get<Iceberg::ManifestFilePtr>(cached_element)->getSizeInMemory() + SIZE_IN_MEMORY_OVERHEAD)
+        , metric_increment{CurrentMetrics::IcebergMetadataFilesCacheSize, memory_bytes}
+    {
+    }
+    IcebergMetadataFilesCacheCell(const IcebergMetadataFilesCacheCell &) = delete;
+    IcebergMetadataFilesCacheCell & operator=(const IcebergMetadataFilesCacheCell &) = delete;
+};
+
+struct IcebergMetadataFilesCacheWeightFunction
+{
+    size_t operator()(const IcebergMetadataFilesCacheCell & cell) const
+    {
+        return cell.memory_bytes;
+    }
+};
+
+class IcebergMetadataFilesCache : public CacheBase<String, IcebergMetadataFilesCacheCell, std::hash<String>, IcebergMetadataFilesCacheWeightFunction>
+{
+public:
+    using Base = CacheBase<String, IcebergMetadataFilesCacheCell, std::hash<String>, IcebergMetadataFilesCacheWeightFunction>;
+
+    IcebergMetadataFilesCache(const String & cache_policy, size_t max_size_in_bytes, size_t max_count, double size_ratio)
+        : Base(cache_policy, max_size_in_bytes, max_count, size_ratio)
+    {}
+
+    static String getKey(StorageObjectStorage::ConfigurationPtr config, const String & data_path)
+    {
+        return std::filesystem::path(config->getDataSourceDescription()) / data_path;
+    }
+
+    template <typename LoadFunc>
+    Poco::JSON::Object::Ptr getOrSetTableMetadata(const String & data_path, LoadFunc && load_fn)
+    {
+        auto load_fn_wrapper = [&]()
+        {
+            const auto & [json_ptr, json_size] = load_fn();
+            return std::make_shared<IcebergMetadataFilesCacheCell>(json_ptr, json_size);
+        };
+        auto result = Base::getOrSet(data_path, load_fn_wrapper);
+        if (result.second)
+            ProfileEvents::increment(ProfileEvents::IcebergMetadataFilesCacheMisses);
+        else
+            ProfileEvents::increment(ProfileEvents::IcebergMetadataFilesCacheHits);
```

Comment:
[just a suggestion to make this code nicer, it's your choice whether to follow:] this part of code from load_fn_wrapper to profile events increment is absolutely identical in these three methods, so I'd move it into a common private templated method.